## CI/CD Failure Analysis

**Analysis Time**: 2026-04-14 06:30 UTC
**Attempt**: #1

### Failed Workflows

| Workflow | Job | Step | Status |
|----------|-----|------|--------|
| Integration Tests | macos-latest - Release | Build | Failed |
| Integration Tests | macos-latest - Debug | Build | Failed |
| Integration Tests | ubuntu-latest - Release | Build | Failed |
| Integration Tests | ubuntu-latest - Debug | Build | Failed |

### Root Cause Analysis

**Primary Error**:
```
tests/unit/test_thread_pool_bridge.cpp:91:9: error: calling a private constructor of class 'ThreadPoolBridge'
tests/unit/test_thread_pool_bridge.cpp:227:21: error: member reference type 'Result<std::shared_ptr<ThreadPoolBridge>>' is not a pointer
```

**Analysis**:
PR #962 migrated `ThreadPoolBridge` to use the Result<T> pattern:
1. Constructor moved to `private` — callers must use `create()` factory method
2. `from_thread_system()` now returns `Result<std::shared_ptr<ThreadPoolBridge>>` instead of raw `std::shared_ptr`
3. `from_common_system()` similarly returns `Result<std::shared_ptr<ThreadPoolBridge>>`

The test file `tests/unit/test_thread_pool_bridge.cpp` was not updated to reflect these API changes.

**Identified Issues**:
1. Direct constructor calls (lines 91, 96) — must use `create()` factory
2. `std::make_shared<ThreadPoolBridge>(...)` usage throughout — must use `create()` factory
3. `from_thread_system()` return value used as raw pointer (lines 224-230) — must unwrap `Result<T>`
4. `ASSERT_NE(bridge, nullptr)` on `Result<T>` — incompatible types

### Proposed Fix

| Issue | Proposed Solution | Files Affected |
|-------|-------------------|----------------|
| Private constructor calls | Replace with `ThreadPoolBridge::create()` + `.value()` | `tests/unit/test_thread_pool_bridge.cpp` |
| `make_shared` usage | Replace with `ThreadPoolBridge::create()` + `.value()` | `tests/unit/test_thread_pool_bridge.cpp` |
| `from_thread_system()` unwrap | Use `ASSERT_TRUE(bridge_result.is_ok())` then `bridge_result.value()` | `tests/unit/test_thread_pool_bridge.cpp` |

### Next Steps
- [ ] Apply proposed fixes to test file
- [ ] Verify locally (if possible)
- [ ] Push and monitor CI

---
*Automated failure analysis - Attempt #1*

--- a/tests/unit/test_dtls_socket.cpp
+++ b/tests/unit/test_dtls_socket.cpp
@@ -101,29 +101,25 @@ protected:
 // Construction Tests
 // ============================================================================
 
-TEST_F(DtlsSocketTest, ConstructWithValidContext)
+TEST_F(DtlsSocketTest, CreateWithValidContext)
 {
 	auto socket = create_server_socket();
-	EXPECT_NO_THROW({
-		auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
-		EXPECT_NE(dtls, nullptr);
-	});
+	auto result = dtls_socket::create(std::move(socket), server_ctx_.get());
+	EXPECT_TRUE(result.is_ok());
+	EXPECT_NE(result.value(), nullptr);
 }
 
-TEST_F(DtlsSocketTest, ConstructWithNullContextThrows)
+TEST_F(DtlsSocketTest, CreateWithNullContextReturnsError)
 {
 	auto socket = create_server_socket();
-	EXPECT_THROW(
-		{
-			auto dtls = std::make_shared<dtls_socket>(std::move(socket), nullptr);
-		},
-		std::runtime_error);
+	auto result = dtls_socket::create(std::move(socket), nullptr);
+	EXPECT_TRUE(result.is_err());
 }
 
 TEST_F(DtlsSocketTest, InitialStateNotHandshakeComplete)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	EXPECT_FALSE(dtls->is_handshake_complete());
 }
@@ -135,7 +131,7 @@ TEST_F(DtlsSocketTest, InitialStateNotHandshakeComplete)
 TEST_F(DtlsSocketTest, SetReceiveCallback)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	std::atomic<bool> callback_set{false};
 
@@ -151,7 +147,7 @@ TEST_F(DtlsSocketTest, SetReceiveCallback)
 TEST_F(DtlsSocketTest, SetErrorCallback)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	EXPECT_NO_THROW({
 		dtls->set_error_callback([](std::error_code) {});
@@ -161,7 +157,7 @@ TEST_F(DtlsSocketTest, SetErrorCallback)
 TEST_F(DtlsSocketTest, SetPeerEndpoint)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	asio::ip::udp::endpoint peer(asio::ip::make_address("127.0.0.1"), 12345);
 	dtls->set_peer_endpoint(peer);
@@ -177,7 +173,7 @@ TEST_F(DtlsSocketTest, SetPeerEndpoint)
 TEST_F(DtlsSocketTest, StartReceiveDoesNotThrow)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	EXPECT_NO_THROW({
 		dtls->start_receive();
@@ -189,7 +185,7 @@ TEST_F(DtlsSocketTest, StartReceiveDoesNotThrow)
 TEST_F(DtlsSocketTest, StopReceiveDoesNotThrow)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	EXPECT_NO_THROW({
 		dtls->stop_receive();
@@ -199,7 +195,7 @@ TEST_F(DtlsSocketTest, StopReceiveDoesNotThrow)
 TEST_F(DtlsSocketTest, MultipleStartReceiveCalls)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	// Multiple start calls should be safe
 	EXPECT_NO_THROW({
@@ -218,7 +214,7 @@ TEST_F(DtlsSocketTest, MultipleStartReceiveCalls)
 TEST_F(DtlsSocketTest, SendBeforeHandshakeFails)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	std::promise<std::error_code> send_promise;
 	auto send_future = send_promise.get_future();
@@ -252,11 +248,11 @@ TEST_F(DtlsSocketTest, ClientServerHandshake)
 {
 	// Create server DTLS socket
 	auto server_udp = create_server_socket();
-	auto server_dtls = std::make_shared<dtls_socket>(std::move(server_udp), server_ctx_.get());
+	auto server_dtls = dtls_socket::create(std::move(server_udp), server_ctx_.get()).value();
 
 	// Create client DTLS socket
 	auto client_udp = create_client_socket();
-	auto client_dtls = std::make_shared<dtls_socket>(std::move(client_udp), client_ctx_.get());
+	auto client_dtls = dtls_socket::create(std::move(client_udp), client_ctx_.get()).value();
 
 	// Set peer endpoints
 	asio::ip::udp::endpoint server_endpoint(asio::ip::make_address("127.0.0.1"), test_port_);
@@ -336,7 +332,7 @@ TEST_F(DtlsSocketTest, ClientServerHandshake)
 TEST_F(DtlsSocketTest, ConcurrentCallbackRegistration)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	std::atomic<int> registration_count{0};
 	const int num_registrations = 100;
@@ -373,7 +369,7 @@ TEST_F(DtlsSocketTest, ConcurrentCallbackRegistration)
 TEST_F(DtlsSocketTest, ConcurrentEndpointAccess)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	std::atomic<int> operation_count{0};
 	const int num_operations = 100;
@@ -419,7 +415,7 @@ TEST_F(DtlsSocketTest, SocketAccessReturnsValidSocket)
 	auto socket = create_server_socket();
 	auto expected_port = socket.local_endpoint().port();
 
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	EXPECT_TRUE(dtls->socket().is_open());
 	EXPECT_EQ(dtls->socket().local_endpoint().port(), expected_port);
@@ -434,7 +430,7 @@ TEST_F(DtlsSocketTest, DestructorAfterStartReceive)
 	// Test that destructor properly cleans up after start_receive
 	EXPECT_NO_THROW({
 		auto socket = create_server_socket();
-		auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+		auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 		dtls->start_receive();
 		// Destructor called when dtls goes out of scope
 	});
@@ -446,7 +442,7 @@ TEST_F(DtlsSocketTest, DestructorAfterStartReceive)
 TEST_F(DtlsSocketTest, MultipleStopReceiveCalls)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	dtls->start_receive();
 
@@ -518,11 +514,11 @@ TEST_F(DtlsSocketIntegrationTest, SendReceiveAfterHandshake)
 {
 	// Create server DTLS socket
 	auto server_udp = create_server_socket();
-	auto server_dtls = std::make_shared<dtls_socket>(std::move(server_udp), server_ctx_.get());
+	auto server_dtls = dtls_socket::create(std::move(server_udp), server_ctx_.get()).value();
 
 	// Create client DTLS socket
 	auto client_udp = create_client_socket();
-	auto client_dtls = std::make_shared<dtls_socket>(std::move(client_udp), client_ctx_.get());
+	auto client_dtls = dtls_socket::create(std::move(client_udp), client_ctx_.get()).value();
 
 	// Set peer endpoints
 	asio::ip::udp::endpoint server_endpoint(asio::ip::make_address("127.0.0.1"), test_port_);
@@ -595,11 +591,11 @@ TEST_F(DtlsSocketIntegrationTest, BidirectionalCommunication)
 {
 	// Create server DTLS socket
 	auto server_udp = create_server_socket();
-	auto server_dtls = std::make_shared<dtls_socket>(std::move(server_udp), server_ctx_.get());
+	auto server_dtls = dtls_socket::create(std::move(server_udp), server_ctx_.get()).value();
 
 	// Create client DTLS socket
 	auto client_udp = create_client_socket();
-	auto client_dtls = std::make_shared<dtls_socket>(std::move(client_udp), client_ctx_.get());
+	auto client_dtls = dtls_socket::create(std::move(client_udp), client_ctx_.get()).value();
 
 	// Set peer endpoints
 	asio::ip::udp::endpoint server_endpoint(asio::ip::make_address("127.0.0.1"), test_port_);
@@ -685,7 +681,7 @@ TEST_F(DtlsSocketIntegrationTest, BidirectionalCommunication)
 TEST_F(DtlsSocketTest, ErrorCallbackInvokedOnSocketClose)
 {
 	auto socket = create_server_socket();
-	auto dtls = std::make_shared<dtls_socket>(std::move(socket), server_ctx_.get());
+	auto dtls = dtls_socket::create(std::move(socket), server_ctx_.get()).value();
 
 	std::promise<std::error_code> error_promise;
 	auto error_future = error_promise.get_future();
@@ -725,11 +721,11 @@ TEST_F(DtlsSocketIntegrationTest, LargePayload)
 {
 	// Create server DTLS socket
 	auto server_udp = create_server_socket();
-	auto server_dtls = std::make_shared<dtls_socket>(std::move(server_udp), server_ctx_.get());
+	auto server_dtls = dtls_socket::create(std::move(server_udp), server_ctx_.get()).value();
 
 	// Create client DTLS socket
 	auto client_udp = create_client_socket();
-	auto client_dtls = std::make_shared<dtls_socket>(std::move(client_udp), client_ctx_.get());
+	auto client_dtls = dtls_socket::create(std::move(client_udp), client_ctx_.get()).value();
 
 	// Set peer endpoints
 	asio::ip::udp::endpoint server_endpoint(asio::ip::make_address("127.0.0.1"), test_port_);

--- a/tests/unit/test_network_system_bridge.cpp
+++ b/tests/unit/test_network_system_bridge.cpp
@@ -78,9 +78,9 @@ class NetworkSystemBridgeTest : public ::testing::Test {
 protected:
     void SetUp() override {
         mock_pool_ = std::make_shared<mock_thread_pool>();
-        thread_pool_bridge_ = std::make_shared<ThreadPoolBridge>(
+        thread_pool_bridge_ = ThreadPoolBridge::create(
             mock_pool_,
-            ThreadPoolBridge::BackendType::Custom);
+            ThreadPoolBridge::BackendType::Custom).value();
     }
 
     void TearDown() override {

--- a/tests/unit/test_observability_bridge.cpp
+++ b/tests/unit/test_observability_bridge.cpp
@@ -135,25 +135,24 @@ protected:
     std::shared_ptr<mock_monitoring> monitor_;
 };
 
-TEST_F(ObservabilityBridgeTest, ConstructorWithNullLoggerThrows) {
-    EXPECT_THROW(
-        ObservabilityBridge(nullptr, monitor_, ObservabilityBridge::BackendType::Standalone),
-        std::invalid_argument);
+TEST_F(ObservabilityBridgeTest, CreateWithNullLoggerReturnsError) {
+    auto result = ObservabilityBridge::create(nullptr, monitor_, ObservabilityBridge::BackendType::Standalone);
+    EXPECT_TRUE(result.is_err());
 }
 
-TEST_F(ObservabilityBridgeTest, ConstructorWithNullMonitorThrows) {
-    EXPECT_THROW(
-        ObservabilityBridge(logger_, nullptr, ObservabilityBridge::BackendType::Standalone),
-        std::invalid_argument);
+TEST_F(ObservabilityBridgeTest, CreateWithNullMonitorReturnsError) {
+    auto result = ObservabilityBridge::create(logger_, nullptr, ObservabilityBridge::BackendType::Standalone);
+    EXPECT_TRUE(result.is_err());
 }
 
-TEST_F(ObservabilityBridgeTest, ConstructorWithValidInterfaces) {
-    EXPECT_NO_THROW(ObservabilityBridge(logger_, monitor_, ObservabilityBridge::BackendType::Standalone));
+TEST_F(ObservabilityBridgeTest, CreateWithValidInterfaces) {
+    auto result = ObservabilityBridge::create(logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    EXPECT_TRUE(result.is_ok());
 }
 
 TEST_F(ObservabilityBridgeTest, InitializeSuccess) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -166,8 +165,8 @@ TEST_F(ObservabilityBridgeTest, InitializeSuccess) {
 }
 
 TEST_F(ObservabilityBridgeTest, InitializeWithMonitoringDisabled) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -182,8 +181,8 @@ TEST_F(ObservabilityBridgeTest, InitializeWithMonitoringDisabled) {
 }
 
 TEST_F(ObservabilityBridgeTest, InitializeDisabledBridgeFails) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -195,8 +194,8 @@ TEST_F(ObservabilityBridgeTest, InitializeDisabledBridgeFails) {
 }
 
 TEST_F(ObservabilityBridgeTest, InitializeAlreadyInitializedFails) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -209,8 +208,8 @@ TEST_F(ObservabilityBridgeTest, InitializeAlreadyInitializedFails) {
 }
 
 TEST_F(ObservabilityBridgeTest, ShutdownSuccess) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -226,8 +225,8 @@ TEST_F(ObservabilityBridgeTest, ShutdownSuccess) {
 }
 
 TEST_F(ObservabilityBridgeTest, ShutdownIdempotent) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -241,8 +240,8 @@ TEST_F(ObservabilityBridgeTest, ShutdownIdempotent) {
 }
 
 TEST_F(ObservabilityBridgeTest, GetLoggerReturnsValidPointer) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -254,8 +253,8 @@ TEST_F(ObservabilityBridgeTest, GetLoggerReturnsValidPointer) {
 }
 
 TEST_F(ObservabilityBridgeTest, GetMonitorReturnsValidPointer) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -267,8 +266,8 @@ TEST_F(ObservabilityBridgeTest, GetMonitorReturnsValidPointer) {
 }
 
 TEST_F(ObservabilityBridgeTest, GetMetricsReturnsCorrectBackendType) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -283,8 +282,8 @@ TEST_F(ObservabilityBridgeTest, GetMetricsReturnsCorrectBackendType) {
 }
 
 TEST_F(ObservabilityBridgeTest, GetMetricsAfterShutdownReportsUnhealthy) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -296,15 +295,15 @@ TEST_F(ObservabilityBridgeTest, GetMetricsAfterShutdownReportsUnhealthy) {
 }
 
 TEST_F(ObservabilityBridgeTest, GetBackendTypeReturnsCorrectType) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     EXPECT_EQ(ObservabilityBridge::BackendType::Standalone, bridge->get_backend_type());
 }
 
 TEST_F(ObservabilityBridgeTest, LoggerUsageAfterInitialization) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -319,8 +318,8 @@ TEST_F(ObservabilityBridgeTest, LoggerUsageAfterInitialization) {
 }
 
 TEST_F(ObservabilityBridgeTest, MonitoringUsageAfterInitialization) {
-    auto bridge = std::make_shared<ObservabilityBridge>(
-        logger_, monitor_, ObservabilityBridge::BackendType::Standalone);
+    auto bridge = ObservabilityBridge::create(
+        logger_, monitor_, ObservabilityBridge::BackendType::Standalone).value();
 
     BridgeConfig config;
     config.integration_name = "test_observability";
@@ -335,14 +334,12 @@ TEST_F(ObservabilityBridgeTest, MonitoringUsageAfterInitialization) {
 }
 
 #if KCENON_WITH_COMMON_SYSTEM
-TEST_F(ObservabilityBridgeTest, FromCommonSystemCreatesValidBridge) {
+TEST_F(ObservabilityBridgeTest, FromCommonSystemWithNullArgsReturnsError) {
     // Note: This test requires common_system to be available
-    // For now, we just verify that the factory method signature exists
-    // and that it would throw with null arguments
+    // For now, we just verify that the factory method returns error with null arguments
 
-    EXPECT_THROW(
-        ObservabilityBridge::from_common_system(nullptr, nullptr),
-        std::invalid_argument);
+    auto result = ObservabilityBridge::from_common_system(nullptr, nullptr);
+    EXPECT_TRUE(result.is_err());
 }
 #endif
 

--- a/tests/unit/test_thread_pool_bridge.cpp
+++ b/tests/unit/test_thread_pool_bridge.cpp
@@ -86,18 +86,18 @@ protected:
     std::shared_ptr<mock_thread_pool> pool_;
 };
 
-TEST_F(ThreadPoolBridgeTest, ConstructorWithNullPoolThrows) {
-    EXPECT_THROW(
-        ThreadPoolBridge(nullptr, ThreadPoolBridge::BackendType::Custom),
-        std::invalid_argument);
+TEST_F(ThreadPoolBridgeTest, CreateWithNullPoolReturnsError) {
+    auto result = ThreadPoolBridge::create(nullptr, ThreadPoolBridge::BackendType::Custom);
+    EXPECT_TRUE(result.is_err());
 }
 
-TEST_F(ThreadPoolBridgeTest, ConstructorWithValidPool) {
-    EXPECT_NO_THROW(ThreadPoolBridge(pool_, ThreadPoolBridge::BackendType::Custom));
+TEST_F(ThreadPoolBridgeTest, CreateWithValidPool) {
+    auto result = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom);
+    EXPECT_TRUE(result.is_ok());
 }
 
 TEST_F(ThreadPoolBridgeTest, InitializeSuccess) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -110,7 +110,7 @@ TEST_F(ThreadPoolBridgeTest, InitializeSuccess) {
 
 TEST_F(ThreadPoolBridgeTest, InitializeWithStoppedPoolFails) {
     pool_->set_running(false);
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -121,7 +121,7 @@ TEST_F(ThreadPoolBridgeTest, InitializeWithStoppedPoolFails) {
 }
 
 TEST_F(ThreadPoolBridgeTest, InitializeDisabledBridgeFails) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -133,7 +133,7 @@ TEST_F(ThreadPoolBridgeTest, InitializeDisabledBridgeFails) {
 }
 
 TEST_F(ThreadPoolBridgeTest, InitializeAlreadyInitializedFails) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -146,7 +146,7 @@ TEST_F(ThreadPoolBridgeTest, InitializeAlreadyInitializedFails) {
 }
 
 TEST_F(ThreadPoolBridgeTest, ShutdownSuccess) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -160,7 +160,7 @@ TEST_F(ThreadPoolBridgeTest, ShutdownSuccess) {
 }
 
 TEST_F(ThreadPoolBridgeTest, ShutdownIdempotent) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -174,14 +174,14 @@ TEST_F(ThreadPoolBridgeTest, ShutdownIdempotent) {
 }
 
 TEST_F(ThreadPoolBridgeTest, GetMetricsBeforeInitialization) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     auto metrics = bridge->get_metrics();
     EXPECT_FALSE(metrics.is_healthy);
 }
 
 TEST_F(ThreadPoolBridgeTest, GetMetricsAfterInitialization) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -195,7 +195,7 @@ TEST_F(ThreadPoolBridgeTest, GetMetricsAfterInitialization) {
 }
 
 TEST_F(ThreadPoolBridgeTest, GetMetricsAfterShutdown) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -207,23 +207,24 @@ TEST_F(ThreadPoolBridgeTest, GetMetricsAfterShutdown) {
 }
 
 TEST_F(ThreadPoolBridgeTest, GetThreadPool) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     auto retrieved_pool = bridge->get_thread_pool();
     EXPECT_EQ(retrieved_pool, pool_);
 }
 
 TEST_F(ThreadPoolBridgeTest, GetBackendType) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(
-        pool_, ThreadPoolBridge::BackendType::ThreadSystem);
+    auto bridge = ThreadPoolBridge::create(
+        pool_, ThreadPoolBridge::BackendType::ThreadSystem).value();
 
     EXPECT_EQ(bridge->get_backend_type(), ThreadPoolBridge::BackendType::ThreadSystem);
 }
 
 TEST_F(ThreadPoolBridgeTest, FromThreadSystemFactoryMethod) {
-    auto bridge = ThreadPoolBridge::from_thread_system("test_pool");
-    ASSERT_NE(bridge, nullptr);
+    auto bridge_result = ThreadPoolBridge::from_thread_system("test_pool");
+    ASSERT_TRUE(bridge_result.is_ok());
 
+    auto bridge = bridge_result.value();
     EXPECT_EQ(bridge->get_backend_type(), ThreadPoolBridge::BackendType::ThreadSystem);
 
     auto pool = bridge->get_thread_pool();
@@ -231,7 +232,7 @@ TEST_F(ThreadPoolBridgeTest, FromThreadSystemFactoryMethod) {
 }
 
 TEST_F(ThreadPoolBridgeTest, ThreadPoolFunctionality) {
-    auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+    auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
     BridgeConfig config;
     config.integration_name = "test_pool";
@@ -251,10 +252,9 @@ TEST_F(ThreadPoolBridgeTest, ThreadPoolFunctionality) {
 
 #if KCENON_WITH_COMMON_SYSTEM
 
-TEST_F(ThreadPoolBridgeTest, FromCommonSystemFactoryMethodWithNullExecutorThrows) {
-    EXPECT_THROW(
-        ThreadPoolBridge::from_common_system(nullptr),
-        std::invalid_argument);
+TEST_F(ThreadPoolBridgeTest, FromCommonSystemFactoryMethodWithNullExecutorReturnsError) {
+    auto result = ThreadPoolBridge::from_common_system(nullptr);
+    EXPECT_TRUE(result.is_err());
 }
 
 // Note: Full test for from_common_system requires a mock IExecutor
@@ -264,7 +264,7 @@ TEST_F(ThreadPoolBridgeTest, FromCommonSystemFactoryMethodWithNullExecutorThrows
 
 TEST_F(ThreadPoolBridgeTest, DestructorCallsShutdownIfInitialized) {
     {
-        auto bridge = std::make_shared<ThreadPoolBridge>(pool_, ThreadPoolBridge::BackendType::Custom);
+        auto bridge = ThreadPoolBridge::create(pool_, ThreadPoolBridge::BackendType::Custom).value();
 
         BridgeConfig config;
         config.integration_name = "test_pool";

